src/main/java/com/twitter/joauth/UrlCodec.java

```diff
@@ -105,21 +105,20 @@ public static String normalize(String s) {
         }
         if (c == '%') {
           if (i + 3 <= length) {
-            // TODO: look into reducing garbage here
-            String spChar = s.substring(i, i + 3).toUpperCase();
-            String toAppend = spChar.toString();
-
-            if (ENCODED_UNDERSCORE.equals(spChar)) {
-              toAppend = UNDERSCORE;
-            } else if (ENCODED_DASH.equals(spChar)) {
-              toAppend = DASH;
-            } else if (ENCODED_TILDE.equals(spChar)) {
-              toAppend = TILDE;
-            } else if (ENCODED_PERIOD.equals(spChar)) {
-              toAppend = PERIOD;
+            if (ENCODED_UNDERSCORE.regionMatches(true, 0, s, i, 3)) {
```

Comment:
We already know the first char in %. So, we have to only match the next two char. It may not make a noticeable difference in perf though. So, I am fine as is.